injector/injector.go

```diff
@@ -0,0 +1,343 @@
+	nonLinearAdsCase   = "NonLinearAds"
+	trackingEventsCase = "TrackingEvents"
+	impressionCase     = "Impression"
+	errorCase          = "Error"
+	companionCase      = "Companion"
+	companionAdsCase   = "CompanionAds"
+)
+
+type Injector interface {
+	InjectTracker(vastXML string, NURL string) string
+}
+
+type VASTEvents struct {
+	Errors                 []string
+	Impressions            []string
+	VideoClicks            []string
+	NonLinearClickTracking []string
+	CompanionClickThrough  []string
+	TrackingEvents         map[string][]string
+}
+
+type InjectionState struct {
+	injectTracker         bool
+	injectVideoClicks     bool
+	inlineWrapperTagFound bool
+	wrapperTagFound       bool
+	impressionTagFound    bool
+	errorTagFound         bool
+	creativeId            string
+	isCreative            bool
+	companionTagFound     bool
+	nonLinearTagFound     bool
+}
+
+type TrackerInjector struct {
+	replacer macros.Replacer
+	events   VASTEvents
+	me       metrics.MetricsEngine
+	provider *macros.MacroProvider
+}
+
+var trimRunes = "\t\r\b\n "
+
+func NewTrackerInjector(replacer macros.Replacer, provider *macros.MacroProvider, events VASTEvents) *TrackerInjector {
+	return &TrackerInjector{
+		replacer: replacer,
+		provider: provider,
+		events:   events,
+	}
+}
+
+func (ti *TrackerInjector) InjectTracker(vastXML string, NURL string) string {
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+	st := &InjectionState{
+		injectTracker:         false,
+		injectVideoClicks:     false,
+		inlineWrapperTagFound: false,
+		wrapperTagFound:       false,
+		impressionTagFound:    false,
+		errorTagFound:         false,
+		creativeId:            "",
+		isCreative:            false,
+		companionTagFound:     false,
+		nonLinearTagFound:     false,
+	}
+
+	reader := strings.NewReader(vastXML)
+	decoder := xml.NewDecoder(reader)
+
+	for {
+		t, err := decoder.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			ti.handleStartElement(tt, st, &outputXML, encoder)
+		case xml.EndElement:
+			ti.handleEndElement(tt, st, &outputXML, encoder)
+		case xml.CharData:
+			tt2 := strings.Trim(string(tt), trimRunes)
+			if len(tt2) != 0 {
+				encoder.Flush()
+				outputXML.WriteString("<![CDATA[")
+				outputXML.WriteString(tt2)
+				outputXML.WriteString("]]>")
+				continue
+			}
+		default:
+			encoder.EncodeToken(t)
+		}
+	}
+
+	encoder.Flush()
+
+	if !st.inlineWrapperTagFound {
+		// Todo log adapter.<bidder-name>.requests.badserverresponse metrics
+		return vastXML
+	}
+	return outputXML.String()
+}
+
+func (ti *TrackerInjector) handleStartElement(tt xml.StartElement, st *InjectionState, outputXML *strings.Builder, encoder *xml.Encoder) {
```

Comment:
Can you explain the variable names `tt` and `st`? I'm wondering if they can be improved.